larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1154 @@
+    const pandora::CartesianVector endPosition(lastTrackState.GetPosition());
+
+    const pandora::CartesianVector lowZVector(initialPosition.GetZ() < endPosition.GetZ() ? initialPosition : endPosition);
+    const pandora::CartesianVector highZVector(initialPosition.GetZ() > endPosition.GetZ() ? initialPosition : endPosition);
+
+    fitResult.SetBeginpoint(lowZVector);
+    fitResult.SetEndpoint(highZVector);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::FillHitChargeVector(const Cluster *const pCluster, HitChargeVector &hitChargeVector)
+{
+    OrderedCaloHitList orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+    CaloHitList caloHitList;
+    orderedCaloHitList.FillCaloHitList(caloHitList);
+
+    const TwoDSlidingFitResult &slidingFit(this->GetCachedSlidingFit(pCluster));
+
+    for (CaloHitList::const_iterator hitIter = caloHitList.begin(), hitIterEnd = caloHitList.end(); hitIter != hitIterEnd; ++hitIter)
+    {
+        const CaloHit *const pCaloHit(*hitIter);
+        const CartesianVector caloHitPosition(pCaloHit->GetPositionVector());
+        float hitWidth(pCaloHit->GetCellSize1());
+
+        float caloHitEnergy(pCaloHit->GetInputEnergy());
+        caloHitEnergy *= 273.5; //ADC to electron
+        caloHitEnergy *= 23.6/1000000; //ionisation energy per electron in MeV
+        caloHitEnergy /= 0.62;
+
+        float rL(0.f), rT(0.f);
+        slidingFit.GetLocalPosition(caloHitPosition, rL, rT);
+        if (rL == 0.)
+            continue;
+
+        float calibratedUncertainty(std::sqrt((0.00419133 * (caloHitEnergy/hitWidth) * (caloHitEnergy/hitWidth)) + (0.00967141 * (caloHitEnergy/hitWidth)))); //70%
+        HitCharge hitCharge(pCaloHit, rL, hitWidth, caloHitEnergy, calibratedUncertainty);
+        hitChargeVector.push_back(hitCharge);
+    }
+
+    std::sort(hitChargeVector.begin(), hitChargeVector.end(), SortHitChargeVectorByRL);
+
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::TrackInnerFilter(HitChargeVector &hitChargeVector, HitChargeVector &filteredHitChargeVector)
+{
+    //Fill endpoint protected area into filtered vector and put all other hits in a separate vector
+    float endpointProtectionRange(0.05);
+    filteredHitChargeVector.insert(filteredHitChargeVector.begin(), hitChargeVector.begin(),  hitChargeVector.begin() + endpointProtectionRange * hitChargeVector.size());
+    filteredHitChargeVector.insert(filteredHitChargeVector.begin(), hitChargeVector.begin() + (1.0 - endpointProtectionRange) * hitChargeVector.size(), hitChargeVector.end());
+
+    HitChargeVector innerHitChargeVector(hitChargeVector.begin() + endpointProtectionRange * hitChargeVector.size(), hitChargeVector.begin() + (1.0 - endpointProtectionRange) * hitChargeVector.size());
```

Comment:
The behaviour is fine as even if this happens on one side, it won't on the other so there's always at least one hit in the filtered vector. 
Do you mean resolving the floats by manually rounding them? Why would this be better than letting the compiler round them?

Reply:
Pointer arithmetic is inherently risky, as it puts the onus on the developer to manage the bounds of any indexed variable, so the compiler won't spot if you've specified an index outside of the bounds of the vector. Exceeding the bounds of a vector is particularly problematic, because you get undefined behaviour, which leads either to spurious results, or to seg faults and bring down an entire production run, for example. As a pointer has to be an integral value, pointer arithmetic using floating point values exacerbates the risk.

In this case, I think the safety and clarity of the code can be improved by assigning the calculated index offsets to integer constants, which can then be checked to ensure the bounds are respected, before using them for the insert operation, for example

`const int offset{static_cast<int>(endpointProtectionRange * hitChargeVector.size())};`
`filteredHitChargeVector.insert(filteredHitChargeVector.begin(), hitChargeVector.begin(),  hitChargeVector.begin() + offset);`

Reply:
Ah right, okay. Done